[Tablet Orders] Present notices across the AdaptiveModalContainer

iamgabrielma opened the pull request:
Work in progress. Closes #12250 

## Description
This PR addresses the problem with `Notices` when using split view, as these appear in one panel only (rather than across the whole `AdaptiveModalContainer`) as well as might display truncated:

<img width="692" alt="Screenshot 2024-03-20 at 14 09 12" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/3795bc5c-c355-48bf-be47-5e17d01c330b">

## Changes
We created a new `Notice` publisher that unifies emissions either from `fixedNotice` or `autodismissableNotice`. When the view model emits a new value for this unified `Notice`, our `AdaptiveModalContainer` will propagate the changes internally and populate it via the `.notice()` view modifier.

## Testing instructions
1. In `EditableOrderViewModel`, add the following code to force Notices:
```
    func forceSyncError() {
        // 1. Test for fixed notice
        fixedNotice = NoticeFactory.createDummyFixedErrorNotice()

        // 2. Test for auto-dismissable notice
        //autodismissableNotice = NoticeFactory.createDummyDissmissableErrorNotice()

        // 3. Test for system error notices
        /*
        orderSynchronizer.statePublisher.map { [weak self] state -> Notice? in
            guard let self = self else { return nil }
            switch state {
            default:
                let error = NSError(domain: "some error", code: 0)
                return NoticeFactory.syncOrderErrorNotice(error, flow: self.flow, with: self.orderSynchronizer)
            }
        }.sink { [weak self] notice in
            guard let self = self else { return }
            if let notice = notice {
                self.fixedNotice = notice
            }
        }
        .store(in: &cancellables)
         */
    }
```
As well as the following dummy Notices in the same file, within `EditableOrderViewModel.NoticeFactory` extension:

```
        static func createDummyFixedErrorNotice() -> Notice {
            return Notice(title: "Error!", subtitle: "Some subtitle", message: "some error message")
        }

        static func createDummyDissmissableErrorNotice() -> Notice {
            return Notice(title: "Error!", subtitle: "Some subtitle", message: "some error message", feedbackType: .error, actionTitle: "Retry?", actionHandler: {
                debugPrint("Action tapped!")
            })
        }
```
Finally, invoke `forceSyncError()` on `EditableOrderViewModel.init()` method.

2. One at a time, comment out the different notice types in `forceSyncError()`, and run the app on an iPad:
```
// 1. Test for fixed notice
// 2. Test for auto-dismissable notice
// 3. Test for system error notices
```
3. Go to `Orders` > `+` > Observe that the notice appears in the bottom of the screen, across the `AdaptiveModalContainer`
4. Observe that Notices work as always on iPhone as well.

## Screenshots
![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-03-20 at 14 04 03](https://github.com/woocommerce/woocommerce-ios/assets/3812076/7a6860c3-568f-4a9a-98a1-bf4fdc87bedb)

<img width="1073" alt="Screenshot 2024-03-20 at 14 45 25" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/b3359865-1581-4abe-9501-bb3618b91d37">

iamgabrielma committed:
Add Notice property to AdaptiveModalContainer. Unify notice types

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -2381,6 +2429,15 @@ extension EditableOrderViewModel {
 extension EditableOrderViewModel {
 
     enum NoticeFactory {
+        static func createDummyFixedErrorNotice() -> Notice {
+            return Notice(title: "Error!", subtitle: "Some subtitle", message: "some error message")
+        }
+
+        static func createDummyDissmissableErrorNotice() -> Notice {
+            return Notice(title: "Error!", subtitle: "Some subtitle", message: "some error message", feedbackType: .error, actionTitle: "Retry?", actionHandler: {
+                debugPrint("Action tapped!")
+            })
+        }
```
Along with the `forceSyncError()` testing function, these will be removed.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/191">17.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Handle duplicated notification on iPhones when flag is active

iamgabrielma committed:
remove debug code

iamgabrielma committed:
Update unit tests

joshheald committed:
12250 Move notice presentation to OrderForm

AdaptiveModalContainer doesn’t need to know anything about notices to do its work (showing two views side by side) since we can show the notices using a modifier at the call site.

joshheald committed:
12250 restore fixed/autodismissable notices

joshheald committed:
12250 Make product notices visible

This avoids the notices being shown under the product screen, when in modal-on-modal mode.

joshheald committed:
Fix crash when syncing variations while offline.

NSError() crashes when called without arguments, and this happened if you sync product variations while offline.

By using a specific swift error, we avoid this crash.

joshheald reviewed (approved):
@iamgabrielma I made quite a lot of changes here. 

I don't think that AdaptiveModalContainer should know about notices – my changes move that to the contained views, and restore the non-autodismiss behaviour for some notices.

Also note that the notices shown by the product selector were shown small and in the left hand pane only – the changes I've added fix that... though I was unable to fix it for Product Variations.

The problem is that notices can't be presented over a sheet, by the view that has the notice modifier... so when we're in modal on modal, we're setting the notices on the contained views.

I don't love this solution either... because I think it needs to be more declarative. The viewModel, via its view, should say "put a notice up", and then the system should (probably) put that atop everything else. That's tricky to do in this PR, but what we should aim for when we work more here.

joshheald committed:
Fix whitespace

iamgabrielma commented:
Thanks for working on this @joshheald, I don't think I was aware of the `.notice` custom modifier. It definitely makes sense to decouple the view from the need to know if there's a notice that should be shown, and using the view modifier rather than injecting the property into the view helps with this and to keep responsibilities separate.

> I don't love this solution either... because I think it needs to be more declarative. The viewModel, via its view, should say "put a notice up", and then the system should (probably) put that atop everything else. That's tricky to do in this PR, but what we should aim for when we work more here.

There are definitely a bunch of instances where we need to tell the system how and when to display those, and would be great to move to a place where the framework decides that based on state only without us having to tweak each scenario. I'm not quite sure how we could improve this with the current design, but I logged a future task so is not lost in the conversation: https://github.com/woocommerce/woocommerce-ios/issues/12389
